crypto/encode_decode/decoder_pkey.c

```
@@ -148,7 +150,25 @@ static int decoder_construct_pkey(OSSL_DECODER_INSTANCE *decoder_inst,
          * result in the keymgmt.
          */
         if (keymgmt_prov == decoder_prov) {
+            /*
+             * When load returns NULL, because, though the provided key material
+             * is syntactically valid (parsed OK), it is not an acceptable key,
+             * the reason why the key is rejected would be lost, unless we take
+             * care to save and restore the errors added once decoding is done.
+             *
+             * The errors are saved in the decoder context, which restores them
+             * when it is cleaned up (freed).
+             */
+            ERR_set_mark();
             keydata = evp_keymgmt_load(keymgmt, object_ref, object_ref_sz);
+            if (keydata == NULL
+                && data->ctx->errkeep == NULL
+                && (data->ctx->errkeep = OSSL_ERR_STATE_new()) != NULL) {
+                OSSL_ERR_STATE_save_to_mark(data->ctx->errkeep);
+                ERR_pop_to_mark();
+            } else {
+                ERR_clear_last_mark();
+            }
```

This is very strange, is the wrong solution and works completely by accident.

Decoding does *not* swallow the errors from load. Calling `OSSL_DECODER_from_bio` (or one of its similarly named friends) will fail with a bad ML-KEM key and the error will still be on the error queue for inspection by the application.

I think the problem you are trying to solve is behaviour in *OSSL_STORE* - not the decoder at all.

Specifically you will note in the store it explicitly and deliberately ignores error codes from `OSSL_DECODER_from_data`:

https://github.com/openssl/openssl/blob/78d252889b7d126b9baf5075db6990658c87f85b/crypto/store/store_result.c#L294-L295

Later, the store will ditch any errors on the error stack and try something else if loading a key failed:

https://github.com/openssl/openssl/blob/78d252889b7d126b9baf5075db6990658c87f85b/crypto/store/store_result.c#L133-L140

Your solution only works because you put an unbalanced `ERR_set_mark` call in `OSSL_DECODER_CTX_free`. Thus when the store calls `ERR_pop_to_mark` it only pops to your unbalanced mark and not the mark that it set itself.

---

I can restore "balance" by calling `ERR_pop_to_mark()` instead of `ERR_clear_error()` before restoring the errors, leaving the mark count unchanged. And then, because load is called twice, the error ends up duplicated:

```
Could not find private key of key from <stdin>
800BBA42067F0000:error:1C80009E:Provider routines:ml_kem_load:invalid key:providers/implementations/keymgmt/ml_kem_kmgmt.c:460:private ML-KEM-512 key implicit rejection secret does not match seed
800BBA42067F0000:error:1C80009E:Provider routines:ml_kem_load:invalid key:providers/implementations/keymgmt/ml_kem_kmgmt.c:460:private ML-KEM-512 key implicit rejection secret does not match seed
800BBA42067F0000:error:1608010C:STORE routines:ossl_store_handle_load_result:unsupported:crypto/store/store_result.c:151:
```

---

Do you have a better solution?

---

Reworked to return the error stack to the store to deal with directly, also now avoids futile second attempt at legacy decoding when the first fails decodes, but fails to load.